utils.py

```
@@ -1,4 +1,49 @@
 import tiktoken
+import subprocess
+import os
+
 def num_tokens_from_string(string: str, encoding_name: str = "cl100k_base") -> int:
     encoding = tiktoken.get_encoding(encoding_name)
-    return len(encoding.encode(string))
\ No newline at end of file
+    return len(encoding.encode(string))
+
+def get_git_commit_hash(repo_path: str = None) -> str:
```

Ruff (0.14.4)

PEP 484 prohibits implicit `Optional`

Convert to `T | None`

(RUF013)